### DISCUSSION TITLE
apollo-openapi: Updating the entire namespace

### DISCUSSION THREAD

**Discussion by Anilople** [apolloconfig/apollo#4640] [General]
current api's document in
* 中文：[Apollo开放平台](https://www.apolloconfig.com/#/zh/usage/apollo-open-api-platform)
* English：[Apollo open-platform](https://www.apolloconfig.com/#/en/usage/apollo-open-api-platform)

Maybe there are some api you want but not in the document, we can disscuss about them, and implement them.

**Comment by Anilople** [comment]
> [3.2.10 新增配置接口](https://www.apolloconfig.com/#/zh/usage/apollo-open-api-platform?id=_3210-%e6%96%b0%e5%a2%9e%e9%85%8d%e7%bd%ae%e6%8e%a5%e5%8f%a3)
> [3.2.11 修改配置接口](https://www.apolloconfig.com/#/zh/usage/apollo-open-api-platform?id=_3211-%e4%bf%ae%e6%94%b9%e9%85%8d%e7%bd%ae%e6%8e%a5%e5%8f%a3)

> [3.2.10 New configuration interface](https://www.apolloconfig.com/#/en/usage/apollo-open-api-platform?id=_3210-new-configuration-interface)
> [3.2.11 Modifying the configuration interface](https://www.apolloconfig.com/#/en/usage/apollo-open-api-platform?id=_3211-modifying-the-configuration-interface)

Maybe we need an interface to modify whole namespace. i.e write multiple key values to namespace once.

For example,

* URL: http://{portal_address}/openapi/v1/envs/{env}/apps/{appId}/clusters/{clusterName}/namespaces/{namespaceName}/items
* Method: PUT

Request Body:
* comment
* dataChangeLastModifiedBy
* dataChangeCreatedBy
* keyValues: a json object

```json
{
  "keyValues": {
    "timeout": "3000",
    "thread": "10"
  },
  "comment":"gateway config",
  "dataChangeLastModifiedBy":"zhanglea"
}
```

**Comment by nobodyiam** [comment]
The Apollo portal offers two types of updates:

1. [Updating a single item](https://github.com/apolloconfig/apollo/blob/3ef13f8b53335b90002d88518c6cb76f294e554e/apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/controller/ItemController.java#L111)
2. [Updating the entire namespace](https://github.com/apolloconfig/apollo/blob/3ef13f8b53335b90002d88518c6cb76f294e554e/apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/controller/ItemController.java#L79)

For apollo openapi, I think it's better to reuse these 2 models for better code maintenance.
